File: backend/routes/onboarding.py
from typing import Any, Optional

from fastapi import APIRouter
from prisma import Json
from pydantic import BaseModel

router = APIRouter()


def get_db():
    from main import db
    return db


class ResponseSave(BaseModel):
    device_id: Optional[str] = None
    user_id: Optional[str] = None
    question_id: str
    answer: Any


@router.get("/questions")
async def get_questions():
    db = get_db()
    questions = await db.question.find_many(
        where={"is_active": True},
        order={"order": "asc"},
    )
    return {"questions": questions}


@router.post("/responses")
async def save_response(body: ResponseSave):
    db = get_db()

    # Look up existing response by user_id first, then device_id
    existing = None
    lookup_key = None
    lookup_value = None

    if body.user_id:
        existing = await db.response.find_unique(where={"user_id": body.user_id})
        lookup_key = "user_id"
        lookup_value = body.user_id
    if not existing and body.device_id:
        existing = await db.response.find_unique(where={"device_id": body.device_id})
        lookup_key = "device_id"
        lookup_value = body.device_id

    if existing:
        current = existing.answers if isinstance(existing.answers, dict) else {}
        current[body.question_id] = body.answer
        response = await db.response.update(
            where={lookup_key: lookup_value},
            data={"answers": Json(current)},
        )
    else:
        # Create new row
        data = {"answers": Json({body.question_id: body.answer})}
        if body.user_id:
            data["user_id"] = body.user_id
        if body.device_id:
            data["device_id"] = body.device_id
        response = await db.response.create(data=data)

    return {"success": True, "response": response}


@router.get("/responses/{identifier}")
async def get_responses(identifier: str):
    db = get_db()
    # Try user_id first, then device_id
    response = await db.response.find_unique(where={"user_id": identifier})
    if not response:
        response = await db.response.find_unique(where={"device_id": identifier})
    total = await db.question.count(where={"is_active": True})
    answers = response.answers if response and isinstance(response.answers, dict) else {}
    answered = len(answers)
    return {
        "response": response,
        "completion": {
            "answered": answered,
            "total": total,
            "percent": round((answered / total) * 100) if total > 0 else 0,
        },
    }

File: backend/src/routes/onboarding.ts
import { Router } from 'express';
import { PrismaClient } from '@prisma/client';

const prisma = new PrismaClient();
export const onboardingRouter = Router();

// GET all active questions (ordered)
onboardingRouter.get('/questions', async (_req, res) => {
  try {
    const questions = await prisma.question.findMany({
      where: { is_active: true },
      orderBy: { order: 'asc' },
    });
    res.json({ questions });
  } catch (err) {
    console.error('Fetch questions error:', err);
    res.status(500).json({ error: 'Failed to fetch questions' });
  }
});

// POST save a single response (upsert)
onboardingRouter.post('/responses', async (req, res) => {
  const { user_id, question_id, answer } = req.body;

  if (!user_id || !question_id || answer === undefined) {
    res.status(400).json({ error: 'user_id, question_id, and answer are required' });
    return;
  }

  try {
    const response = await prisma.response.upsert({
      where: {
        user_id_question_id: { user_id, question_id },
      },
      update: { answer },
      create: { user_id, question_id, answer },
    });
    res.json({ success: true, response });
  } catch (err) {
    res.status(500).json({ error: 'Failed to save response' });
  }
});

// GET all responses for a user + completion stats
onboardingRouter.get('/responses/:userId', async (req, res) => {
  try {
    const totalQuestions = await prisma.question.count({
      where: { is_active: true },
    });

    const responses = await prisma.response.findMany({
      where: { user_id: req.params.userId },
      include: { question: true },
      orderBy: { question: { order: 'asc' } },
    });

    res.json({
      responses,
      completion: {
        answered: responses.length,
        total: totalQuestions,
        percent: totalQuestions > 0 ? Math.round((responses.length / totalQuestions) * 100) : 0,
      },
    });
  } catch (err) {
    res.status(500).json({ error: 'Failed to fetch responses' });
  }
});
